### PR TITLE
Add skip initial table publish with kafka

### LINKED ISSUE #4442: Expose ability to skip initial table publish with kafka.
As a engineer working on deephaven, I would like to have the ability to configure whether or not the initial table state sent to a kafka publisher is published. 

The use case for this is that I have a local state file that I have configured Deephaven to leverage for storing intermediate aggregation states, and persisting them through restarts. This works flawlessly, however on startup currently emits every row from the saved state, this is unnecessary and just adds considerable load to both Kafka and our downstream services to reprocess already emitted data. Ideally we would have a toggle that we could pass the publisher to skip the initial table state sync.

### PR DESCRIPTION
**Pull request by devinrsmith**
Adds builder object KafkaPublishOptions

Fixes #4442

### CODE REVIEW

**Review by rcaudy — APPROVED** [review]

File: py/server/deephaven/stream/kafka/producer.py [hunk 1]
```diff
@@ -32,12 +34,13 @@ def j_object(self) -> jpy.JType:
...
-    key_spec: KeyValueSpec,
-    value_spec: KeyValueSpec,
-    last_by_key_columns: bool = False,
+        table: Table,
+        kafka_config: Dict,
+        topic: str,
+        key_spec: KeyValueSpec,
+        value_spec: KeyValueSpec,
+        last_by_key_columns: bool = False,
+        publish_initial: bool = True,
```
Comment (jmao-denver) [inline py/server/deephaven/stream/kafka/producer.py:43]: inconsistent indentation?

File: py/server/deephaven/stream/kafka/producer.py [hunk 2]
```diff
@@ -70,22 +75,28 @@ def produce(
...
+            runnable = _JKafkaTools.produceFromTable(options)
+
         def cleanup():
             try:
                 runnable.run()
             except Exception as ex:
-                raise DHError(ex, "failed to stop publishing to Kafka and the clean-up.") from ex
+                raise DHError(
+                    ex, "failed to stop publishing to Kafka and the clean-up."
+                ) from ex
```
Comment (jmao-denver) [inline py/server/deephaven/stream/kafka/producer.py:99]: not sure why it needs to be multiple lines.

File: py/server/deephaven/stream/kafka/producer.py [hunk 3]
```diff
@@ -56,6 +59,8 @@ def produce(
         last_by_key_columns (bool): whether to publish only the last record for each unique key, Ignored if key_spec is
             KeyValueSpec.IGNORE. Otherwise, if last_by_key_columns is true this method will internally perform a last_by
             aggregation on table grouped by the input columns of key_spec and publish to Kafka from the result.
+        publish_initial (bool): whether the initial data in table should be published. When False, table.is_refreshing
+            must be True. By default, is True.
```
Comment (jmao-denver) [inline py/server/deephaven/stream/kafka/producer.py:63]: Add a check for the case of  `publish_initial = False and not table.is_refreshing`?

**Review by jmao-denver — CHANGES_REQUESTED** [review]

File: py/server/deephaven/stream/kafka/producer.py (hunk 2 above)
Comment (devinrsmith) [reply py/server/deephaven/stream/kafka/producer.py:99]: I used black formatter; but until we apply it everywhere, I'll revert this part and make a single line.

File: py/server/deephaven/stream/kafka/producer.py [hunk 4]
```diff
@@ -70,17 +75,23 @@ def produce(
...
-        kafka_config = j_properties(kafka_config)
-        runnable = _JKafkaTools.produceFromTable(
-            table.j_table,
-            kafka_config,
-            topic,
-            key_spec.j_object,
-            value_spec.j_object,
-            last_by_key_columns,
+        if not publish_initial and not table.is_refreshing:
+            raise ValueError("publishInitial == False and table.is_refreshing == False")
```
Comment (jmao-denver) [inline py/server/deephaven/stream/kafka/producer.py:79]:
```suggestion
            raise ValueError("publish_initial == False and table.is_refreshing == False")
```

**Review by jmao-denver — APPROVED** [review]
LGTM

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/3151
Reference: https://github.com/deephaven/deephaven.io/issues/3152

**Pull request merged**